SAPT(DFT) External Potential

Awallace3 opened the pull request:
## Description
<!-- Provide a brief description of the PR's purpose here. -->
The objective of this PR is to enable external potentials to be used for both SAPT(DFT) and SAPT(HF).

## User API & Changelog headlines
<!-- A bullet-point format description of how this PR affects the user.
     This is destined for the release notes. May be empty. -->
- [ ] Enables users to call `psi4.energy('sapt(dft)', external_potenial=<EXT_POTENTIAL>)` and acquire SAPT(HF) and SAPT(DFT) energies in the presence of an external potential.

## Dev notes & details
<!-- A bullet-point format description of what this PR does "at a glance."
     Target audience is code reviewers and other devs skimming PRs.
     Should be more technical than user notes. Should never be empty. -->
- [ ] The present implementation leverages the `scf_helper()` external_potential logic to generate and re-use the jk-object to obtain the correct SAPT external potential energies; however, this does not re-use the logic of FISAPT's external potential code. Another ongoing implementation of external potentials in SAPT(DFT) can be found on [this branch](https://github.com/Awallace3/psi4/tree/ext_pot_saptdft) for long term goals of migrating FI-SAPT functionality into SAPT(DFT).
- [ ] This PR depends on [PR#3256](https://github.com/psi4/psi4/pull/3256) for re-using jk-objects and [PR#3232](https://github.com/psi4/psi4/pull/3232) for automatic GRAC shifts. 

## Questions
- [ ] Do we require any additional information about the external potential energies for SAPT(HF) and SAPT(DFT) that isn't provided from the `scf_helper()` route of re-using jk-objects? Most of the logic in FISAPT0 that is neglected is for handling FI-SAPT. 

## Checklist
- [X] Tests added for any new features (see `./tests/pytests/test_saptdft.py` with specifically `test_sapthf_external_potential()` corresponding to the FISAPT0 test located `psi4/tests/fsapt-ext/input.dat`.
- [ ] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [ ] Ready for review
- [ ] Ready for merge

JonathonMisiewicz commented:
Rebase when #3256 is merged in, and then ping me for review.

Awallace3 commented:
> Rebase when #3256 and #3232 are merged in, and then ping me for review.

Will do! Thank you

JonathonMisiewicz commented:
If you want tests to pass, you'll need to borrow [this line](https://github.com/psi4/psi4/pull/3272/commits/d0e24066fd91f1b730907c1c5357640aa0f92c4a#diff-52438ad23e15663afeaef25cc2a0b5e21b154e99149a6a12a63668d8ba27e016R268) from one of Lori's PRs.

JonathonMisiewicz commented:
Also, please go back and _actually_ fix the merge conflict. The `.gitignore` file alone tells me that needs to be re-checked.

Awallace3 commented:
> Also, please go back and _actually_ fix the merge conflict. The `.gitignore` file alone tells me that needs to be re-checked.

I handled the merge conflict in sapt_proc.py; however, I overlooked the .gitignore conflict. Looking back at this branch from your comment, I realized I never cleaned up fisapt.cc to remove debugging related changes, so I have corrected that as well now.

JonathonMisiewicz commented:
SAPT DFT tests aren't passing. Please fix.

Awallace3 commented:
> SAPT DFT tests aren't passing. Please fix.

I have updated the tests and added an explicit test for SAPT(HF) == FISAPT0 for the external potential case ABC.
